src/nvim/syntax.c

```diff
@@ -7227,61 +7297,73 @@ static int syn_add_group(char_u *name)
   if (highlight_ga.ga_len >= MAX_HL_ID) {
     EMSG(_("E849: Too many highlight and syntax groups"));
     xfree(name);
-    return 0;
+    return invalid_group_id;
   }
 
   // Append another syntax_highlight entry.
   struct hl_group* hlgp = GA_APPEND_VIA_PTR(struct hl_group, &highlight_ga);
   memset(hlgp, 0, sizeof(*hlgp));
   hlgp->sg_name = name;
   hlgp->sg_name_u = vim_strsave_up(name);
+  hlgp->sg_link = invalid_group_id;
 
-  return highlight_ga.ga_len;               /* ID is index plus one */
+  // ILOG("Adding group name %s as id %d", name, highlight_ga.ga_len-1);
+
+  return highlight_ga.ga_len-1;
 }
 
-/*
- * When, just after calling syn_add_group(), an error is discovered, this
- * function deletes the new name.
- */
+/// When, just after calling syn_add_group(), an error is discovered, this
+/// function deletes the new name.
 static void syn_unadd_group(void)
 {
-  --highlight_ga.ga_len;
+  highlight_ga.ga_len--;
   xfree(HL_TABLE()[highlight_ga.ga_len].sg_name);
   xfree(HL_TABLE()[highlight_ga.ga_len].sg_name_u);
 }
 
-/*
- * Translate a group ID to highlight attributes.
- */
+
+/// Translate a group ID to highlight attributes.
+/// @see syn_cterm_attr2entry
 int syn_id2attr(int hl_id)
 {
   struct hl_group     *sgp;
 
   hl_id = syn_get_final_id(hl_id);
-  sgp = &HL_TABLE()[hl_id - 1];             /* index is ID minus one */
+  sgp = &HL_TABLE()[hl_id];
   return sgp->sg_attr;
 }
 
+bool hl_is_valid(int hl_id)
+{
+  return !hl_invalid_id(hl_id);
+}
 
-/*
- * Translate a group ID to the final group ID (following links).
- */
+bool hl_invalid_id(int hl_id)
```

Comment:
I added the FUNC_ATTR_PURE. I renamed hl_is_valid to hl_valid_id. I think it makes sense to keep one function available (either hl_invalid_id or hl_valid_id) for other units for instance it is used in tui.c. I can inline or remove the 2nd one.